### PR TITLE
[SPARK-38888][BUILD][CORE][YARN][DOCS] Add `RocksDB` support for shuffle service state store

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
This is a extended work of SPARK-38909, in this pr, the `RocksDB` implementation is added for shuffle local state store.

This PR adds the following code:
- `shuffledb.RocksDB` and `shuffledb.RocksDBIterator`: implementation of RocksDB corresponding to `shuffledb.DB` and `shuffledb.DBIterator`
- Add `ROCKSDB` to shuffle.DBBackend and the corresponding file suffix is `.rdb` and the description of `SHUFFLE_SERVICE_DB_BACKEND` in also changed
- Add `RocksDBProvider` to build `RocksDB` instance and extend `DBProvider` to produce corresponding instances
- Add dependency of `rocksdbjni` to `network-common` module

### Why are the changes needed?
Support shuffle local state store to use RocksDB

### Does this PR introduce _any_ user-facing change?
When user configures `spark.shuffle.service.db.enabled` as true, the user can use rocksdb as the shuffle lcoal state store by specifying `SHUFFLE_SERVICE_DB_BACKEND(spark.shuffle.service.db.backend)` as `RocksDB` in `spark-default.conf` or `spark-shuffle-site.xml(for yarn)`.

The original data store in `LevelDB/RocksDB` will not be automatically convert to another kind of storage now.

### How was this patch tested?
Add new test.

### CODE REVIEW

**Comment by LuciferYang** [comment]
Testing first and will update the pr description later

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (LuciferYang) [inline common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:805]: @zhouyejoe should we set `db = null`?
I found that `Finalized merged shuffle are written into DB and cleaned up after application stopped` in `YarnShuffleServiceWithRocksDBBackendSuite` will crash when use RocksDB.

The native stack as follows:

```
Stack: [0x0000000288410000,0x0000000288813000],  sp=0x0000000288812150,  free space=4104k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
C  [librocksdbjni6224075543374471149.jnilib+0xd4ec8]  _ZN7rocksdb15CompactedDBImpl3GetERKNS_11ReadOptionsEPNS_18ColumnFamilyHandleERKNS_5SliceEPNS_13PinnableSliceEPNSt3__112basic_stringIcNSB_11char_traitsIcEENSB_9allocatorIcEEEE+0xc6c
C  [librocksdbjni6224075543374471149.jnilib+0x1406a0]  _ZN7rocksdb6DBImpl6DeleteERKNS_12WriteOptionsEPNS_18ColumnFamilyHandleERKNS_5SliceE+0x38
C  [librocksdbjni6224075543374471149.jnilib+0x1e38c]  _Z21rocksdb_delete_helperP7JNIEnv_PN7rocksdb2DBERKNS1_12WriteOptionsEPNS1_18ColumnFamilyHandleEP11_jbyteArrayii+0xe0
j  org.rocksdb.RocksDB.delete(J[BII)V+0
j  org.rocksdb.RocksDB.delete([B)V+9
j  org.apache.spark.network.shuffledb.RocksDB.delete([B)V+5
j  org.apache.spark.network.shuffle.RemoteBlockPushResolver.removeAppShufflePartitionInfoFromDB(Lorg/apache/spark/network/shuffle/RemoteBlockPushResolver$AppAttemptShuffleMergeId;)V+16
j  org.apache.spark.network.shuffle.RemoteBlockPushResolver.lambda$removeAppShuffleInfoFromDB$8(Lorg/apache/spark/network/shuffle/RemoteBlockPushResolver$AppShuffleInfo;Ljava/lang/Integer;Lorg/apache/spark/network/shuffle/RemoteBlockPushResolver$AppShuffleMergePartitionsInfo;)V+24
j  org.apache.spark.network.shuffle.RemoteBlockPushResolver$$Lambda$321.accept(Ljava/lang/Object;Ljava/lang/Object;)V+16
j  java.util.concurrent.ConcurrentHashMap.forEach(Ljava/util/function/BiConsumer;)V+56
j  org.apache.spark.network.shuffle.RemoteBlockPushResolver.removeAppShuffleInfoFromDB(Lorg/apache/spark/network/shuffle/RemoteBlockPushResolver$AppShuffleInfo;)V+18
j  org.apache.spark.network.shuffle.RemoteBlockPushResolver.closeAndDeletePartitionsIfNeeded(Lorg/apache/spark/network/shuffle/RemoteBlockPushResolver$AppShuffleInfo;Z)V+25
j  org.apache.spark.network.shuffle.RemoteBlockPushResolver.lambda$applicationRemoved$5(Lorg/apache/spark/network/shuffle/RemoteBlockPushResolver$AppShuffleInfo;Z)V+3
j  org.apache.spark.network.shuffle.RemoteBlockPushResolver$$Lambda$319.run()V+12
j  java.util.concurrent.ThreadPoolExecutor.runWorker(Ljava/util/concurrent/ThreadPoolExecutor$Worker;)V+95
j  java.util.concurrent.ThreadPoolExecutor$Worker.run()V+5
j  io.netty.util.concurrent.FastThreadLocalRunnable.run()V+4
j  java.lang.Thread.run()V+11
v  ~StubRoutines::call_stub
V  [libjvm.dylib+0x289d70]
V  [libjvm.dylib+0x2887fc]
V  [libjvm.dylib+0x288a50]
V  [libjvm.dylib+0x3118d0]
V  [libjvm.dylib+0x524408]
V  [libjvm.dylib+0x5242a4]
V  [libjvm.dylib+0x456874]
C  [libsystem_pthread.dylib+0x7878]  _pthread_start+0x140
C  [libsystem_pthread.dylib+0x25e0]  thread_start+0x8

Java frames: (J=compiled Java code, j=interpreted, Vv=VM code)
j  org.rocksdb.RocksDB.delete(J[BII)V+0
j  org.rocksdb.RocksDB.delete([B)V+9
j  org.apache.spark.network.shuffledb.RocksDB.delete([B)V+5
j  org.apache.spark.network.shuffle.RemoteBlockPushResolver.removeAppShufflePartitionInfoFromDB(Lorg/apache/spark/network/shuffle/RemoteBlockPushResolver$AppAttemptShuffleMergeId;)V+16
j  org.apache.spark.network.shuffle.RemoteBlockPushResolver.lambda$removeAppShuffleInfoFromDB$8(Lorg/apache/spark/network/shuffle/RemoteBlockPushResolver$AppShuffleInfo;Ljava/lang/Integer;Lorg/apache/spark/network/shuffle/RemoteBlockPushResolver$AppShuffleMergePartitionsInfo;)V+24
j  org.apache.spark.network.shuffle.RemoteBlockPushResolver$$Lambda$321.accept(Ljava/lang/Object;Ljava/lang/Object;)V+16
j  java.util.concurrent.ConcurrentHashMap.forEach(Ljava/util/function/BiConsumer;)V+56
j  org.apache.spark.network.shuffle.RemoteBlockPushResolver.removeAppShuffleInfoFromDB(Lorg/apache/spark/network/shuffle/RemoteBlockPushResolver$AppShuffleInfo;)V+18
j  org.apache.spark.network.shuffle.RemoteBlockPushResolver.closeAndDeletePartitionsIfNeeded(Lorg/apache/spark/network/shuffle/RemoteBlockPushResolver$AppShuffleInfo;Z)V+25
j  org.apache.spark.network.shuffle.RemoteBlockPushResolver.lambda$applicationRemoved$5(Lorg/apache/spark/network/shuffle/RemoteBlockPushResolver$AppShuffleInfo;Z)V+3
j  org.apache.spark.network.shuffle.RemoteBlockPushResolver$$Lambda$319.run()V+12
j  java.util.concurrent.ThreadPoolExecutor.runWorker(Ljava/util/concurrent/ThreadPoolExecutor$Worker;)V+95
j  java.util.concurrent.ThreadPoolExecutor$Worker.run()V+5
j  io.netty.util.concurrent.FastThreadLocalRunnable.run()V+4
j  java.lang.Thread.run()V+11
v  ~StubRoutines::call_stub
```
[hs_err_pid12358.log](https://github.com/apache/spark/files/9392790/hs_err_pid12358.log)

We may have operated a closed DB due to `close` and 'removeappshufflepartitioninfofromdb' are on different threads,

Comment (LuciferYang) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:805]: After thinking about it, set `db=null` may not be enough although the test can be work around,  may should use a lock.

File: common/network-common/src/main/java/org/apache/spark/network/util/RocksDBProvider.java
Comment (LuciferYang) [inline common/network-common/src/main/java/org/apache/spark/network/util/RocksDBProvider.java:43]: cc @dongjoon-hyun Do you have time to help review the RocksDB part?

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (mridulm) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:805]: We should be shutting down `mergedShuffleCleaner` executor and waiting for its completion before closing the db.

```
  public void close() {
    if (! mergedShuffleCleaner.isShutdown()) {
      mergedShuffleCleaner.shutdown();
      mergedShuffleCleaner.awaitTermination( ... );
    }
    if (db != null) {
     ...
```

File: docs/spark-standalone.md
Comment (mridulm) [inline docs/spark-standalone.md:341]: Where is this being used ?
Also, `.enabled` typically indicates boolean

**Review by mridulm — COMMENTED** [review]
Mostly looks good, but for a couple of comments.
Particularly have not looked at `RocksDBProvider` - will let @dongjoon-hyun review that (in addition to rest of the PR).

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (LuciferYang) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java:805]: hmm... should we fix this issue in this pr? I think it looks more like a bug fix, so should we use an independent jira to tracking it?

File: docs/spark-standalone.md
Comment (LuciferYang) [reply docs/spark-standalone.md:341]: Sorry, it's my fault. The name of this config should be `spark.shuffle.service.db.backend`.

File: common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java
Comment (LuciferYang) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java]: @mlaflamm @zhouyejoe create SPARK-40186 to tracking this and ignore `Finalized merged shuffle are written into DB and cleaned up after application stopped` in this pr. 

We should re-enable the test after SPARK-40186  resolved and if If SPARK-40186 is resolved before the current pr I will merge it

Comment (LuciferYang) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java]: An independent pr try to solves this: https://github.com/apache/spark/pull/37624

Comment (mridulm) [reply common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java]: Thanks for following up on this @LuciferYang !

File: docs/spark-standalone.md
Comment (mridulm) [reply docs/spark-standalone.md:341]: Should this be in spark-standalone ? applies to others as well, right ? (yarn for example ?)

Comment (LuciferYang) [reply docs/spark-standalone.md:341]: Yes, you are right ~ and `spark.shuffle.service.db.enabled` be used together

https://github.com/apache/spark/blob/a12d121159d0ab8293f70b819cb489cf6126224d/docs/spark-standalone.md?plain=1#L321-L328

Comment (mridulm) [reply docs/spark-standalone.md:341]: What I meant was, add this to 'External Shuffle service(server) side configuration options' section of configuration.md as well ?

Comment (LuciferYang) [reply docs/spark-standalone.md:341]: Got it.

Now `spark.shuffle.service.enabled` is in `configuration.md`, 
`spark.shuffle.service.db.enabled`and `spark.shuffle.service.db.backend` are in `spark-standalone.md`, 
should we add  both`spark.shuffle.service.db.enabled`and `spark.shuffle.service.db.backend` in  `configuration.md`?

Comment (LuciferYang) [reply docs/spark-standalone.md:341]: new diff is https://github.com/apache/spark/compare/db90b273fb941e9f32010562f91a644777605825..15b23255f955ef746fdd292b3fba67dff6c4cd4f

I copy both `spark.shuffle.service.db.enabledand` and  `spark.shuffle.service.db.backend` to `configuration.md `

Comment (mridulm) [reply docs/spark-standalone.md:341]: Yes, both - thanks

**Comment by LuciferYang** [comment]
friendly ping @Ngone51

File: docs/configuration.md
Comment (mridulm) [inline docs/configuration.md:1020]: This should be in the section at the bottom - in `External Shuffle service(server) side configuration options` section.

Having said that, I was wrong about `spark.shuffle.service.db.enabled` - it is always enabled in yarn mode - so we cannot control it with `.enabled` flag.
The newly introduced `spark.shuffle.service.db.backend` is relevant though; but can we add a blurb that it is relevant for yarn and standalone - with db enabled by default for yarn, and to look at standalone.md for more on details there ?

Currently, we have the yarn shuffle service config in `configuration.md`; so can we

Comment (LuciferYang) [reply docs/configuration.md]: Does [e124c84](https://github.com/apache/spark/pull/37610/commits/e124c845eee9c534dff913baa25c3b5a65c83113)  look better?

Comment (LuciferYang) [reply docs/configuration.md]: [4d5f44c](https://github.com/apache/spark/pull/37610/commits/4d5f44ce0b924e8f52f787843ad4759a1731dc4a) move the to bottom of this section

Comment (LuciferYang) [reply docs/configuration.md]: https://spark.apache.org/docs/latest/configuration.html#shuffle-behavior

Is this table item should sorted by `Property Name`? Seems not follow this rule.

**Comment by HyukjinKwon** [comment]
cc @HeartSaVioR FYI

**Comment by LuciferYang** [comment]
thanks @HyukjinKwon

File: docs/configuration.md
Comment (Ngone51) [inline docs/configuration.md:1109]:
```suggestion
    To determine whether external shuffle service uses db to store shuffle state.
```

File: docs/configuration.md
Comment (Ngone51) [inline docs/configuration.md:1111]: Should also link the detailed page for Yarn?

File: docs/configuration.md
Comment (Ngone51) [inline docs/configuration.md:1122]: ditto

File: docs/spark-standalone.md
Comment (Ngone51) [inline docs/spark-standalone.md:337]: Do you mean yarn always enable db or always use LEVELDB as its db? Could you make it more clear?

File: docs/spark-standalone.md
Comment (Ngone51) [inline docs/spark-standalone.md:338]: Do you mean the data that is stored in LevelDB is not available after changing to RocksDB for example?

File: common/network-common/src/main/java/org/apache/spark/network/shuffledb/RocksDBIterator.java
Comment (Ngone51) [inline common/network-common/src/main/java/org/apache/spark/network/shuffledb/RocksDBIterator.java:87]: Wondering why we have a `remove()` (and even no parameter) API for the iterator? What's the expected behavior with a possible implementation?

File: common/network-common/src/main/java/org/apache/spark/network/util/RocksDBProvider.java
Comment (Ngone51) [inline common/network-common/src/main/java/org/apache/spark/network/util/RocksDBProvider.java:81]: Is it possible to have sub-dir here?

File: common/network-common/src/main/java/org/apache/spark/network/util/RocksDBProvider.java
Comment (Ngone51) [inline common/network-common/src/main/java/org/apache/spark/network/util/RocksDBProvider.java:90]: This will not take effect if `dbFile.delete()` failed?

File: common/network-common/src/main/java/org/apache/spark/network/util/RocksDBProvider.java
Comment (Ngone51) [inline common/network-common/src/main/java/org/apache/spark/network/util/RocksDBProvider.java:125]: Why don't we have `setCreateIfMissing` resetting as we do in ` initRockDB(File dbFile, StoreVersion version, ObjectMapper mapper)`? Is the call on this method supposed to be the first time RocksDB creation?

File: docs/spark-standalone.md
Comment (LuciferYang) [reply docs/spark-standalone.md:338]: Yes, there is no function to automatically switch `*.ldb` to `*.rdb` now.

File: common/network-common/src/main/java/org/apache/spark/network/shuffledb/RocksDBIterator.java
Comment (LuciferYang) [reply common/network-common/src/main/java/org/apache/spark/network/shuffledb/RocksDBIterator.java:87]: will remove this , I found 
```
default void remove() {
        throw new UnsupportedOperationException("remove");
    }
```

in java.util.Iterator

File: docs/configuration.md
Comment (LuciferYang) [reply docs/configuration.md:1122]: Is it appropriate to add `spark.shuffle.service.db.backend` to `running-on-yarn.md`? The configurations listed in `running-on-yarn.md` are all in `spark.yarn.shuffle` namespace now.

File: docs/configuration.md
Comment (LuciferYang) [reply docs/configuration.md:1111]: No, yarn always has this behavior `spark.shuffle.service.db.enabled` enabled, I will remove `and Yarn modes`

File: docs/configuration.md
Comment (LuciferYang) [reply docs/configuration.md:1122]: If appropriate, I will add `spark.shuffle.service.db.backend`  to `running-on-yarn.md` and link it

File: docs/spark-standalone.md
Comment (LuciferYang) [reply docs/spark-standalone.md:337]: will delete this

File: common/network-common/src/main/java/org/apache/spark/network/util/RocksDBProvider.java
Comment (LuciferYang) [reply common/network-common/src/main/java/org/apache/spark/network/util/RocksDBProvider.java:81]: The process here refers to LevelDBProvider:

https://github.com/apache/spark/blob/a27238a0df642ec16b31e854801671999f2b76cc/common/network-common/src/main/java/org/apache/spark/network/util/LevelDBProvider.java#L40-L85

I think it is safer to follow the old code flow, could `LevelDBProvider` and `RocksDBProvider` be fixed together after further investigation?

File: common/network-common/src/main/java/org/apache/spark/network/util/RocksDBProvider.java
Comment (LuciferYang) [reply common/network-common/src/main/java/org/apache/spark/network/util/RocksDBProvider.java:125]: Line 58 set `dbOptions.setCreateIfMissing(false)` first and want to restore an existing RocksDB instance.

If the restore catch `RocksDBException`,  set  `dbOptions.setCreateIfMissing(true)` enter the process of creating a new RocksDB instance.

File: common/network-common/src/main/java/org/apache/spark/network/util/RocksDBProvider.java
Comment (LuciferYang) [reply common/network-common/src/main/java/org/apache/spark/network/util/RocksDBProvider.java:81]: I thin this scenario here is a directory or file with the same name as `dbFile`,  but it cannot be recovered, and it is not in the `NotFound` state

Comment (LuciferYang) [reply common/network-common/src/main/java/org/apache/spark/network/util/RocksDBProvider.java:81]: For it is a directory, it looks like a corner case

File: common/network-common/src/main/java/org/apache/spark/network/util/RocksDBProvider.java
Comment (LuciferYang) [reply common/network-common/src/main/java/org/apache/spark/network/util/RocksDBProvider.java:90]: If `dbFile.delete()` fails, the following `RocksDB.open` will also fail

File: docs/configuration.md
Comment (LuciferYang) [reply docs/configuration.md:1111]: done

File: common/network-common/src/main/java/org/apache/spark/network/shuffledb/RocksDBIterator.java
Comment (LuciferYang) [reply common/network-common/src/main/java/org/apache/spark/network/shuffledb/RocksDBIterator.java]: done

File: common/network-common/src/main/java/org/apache/spark/network/shuffledb/RocksDBIterator.java
Comment (LuciferYang) [reply common/network-common/src/main/java/org/apache/spark/network/shuffledb/RocksDBIterator.java]: done

File: common/network-common/src/main/java/org/apache/spark/network/shuffledb/RocksDBIterator.java
Comment (LuciferYang) [reply common/network-common/src/main/java/org/apache/spark/network/shuffledb/RocksDBIterator.java]: done

File: common/network-common/src/main/java/org/apache/spark/network/shuffledb/RocksDBIterator.java
Comment (LuciferYang) [reply common/network-common/src/main/java/org/apache/spark/network/shuffledb/RocksDBIterator.java]: done

File: docs/spark-standalone.md
Comment (LuciferYang) [reply docs/spark-standalone.md]: deleted

File: docs/configuration.md
Comment (LuciferYang) [reply docs/configuration.md:1122]: is [0920f0e](https://github.com/apache/spark/pull/37610/commits/0920f0e000e638f15d283e750b639d41b3501fab) ok ?

File: common/network-common/src/main/java/org/apache/spark/network/util/RocksDBProvider.java
Comment (LuciferYang) [reply common/network-common/src/main/java/org/apache/spark/network/util/RocksDBProvider.java:125]: Sorry, this method just use by test

Comment (LuciferYang) [reply common/network-common/src/main/java/org/apache/spark/network/util/RocksDBProvider.java:125]: https://github.com/apache/spark/blob/43b02d7f25b2ab99f9b0be3c45aaedb1a8bdd40c/resource-managers/yarn/src/test/scala/org/apache/spark/network/shuffle/ShuffleTestAccessor.scala#L210-L217

Comment (LuciferYang) [reply common/network-common/src/main/java/org/apache/spark/network/util/RocksDBProvider.java:125]: move to a new Test Utils?

File: docs/configuration.md
Comment (mridulm) [reply docs/configuration.md]: Actually, that section I mentioned is for push based shuffle - so probably does not make sense for our changes here.

Comment (LuciferYang) [reply docs/configuration.md]: Move it back to after https://github.com/apache/spark/blob/6e62b93f3d1ef7e2d6be0a3bb729ab9b2d55a36d/docs/configuration.md?plain=1#L1000-L1008
?

They have been add into `spark-standalone.md` and `running-on-yarn.md`.Is it necessary to keep them in `configuration.md`?

Comment (mridulm) [reply docs/configuration.md]: You are right, we need it only in standalone and yarn docs for now - not in configuration

Comment (LuciferYang) [reply docs/configuration.md]: removed

File: common/network-common/src/main/java/org/apache/spark/network/util/RocksDBProvider.java
Comment (LuciferYang) [reply common/network-common/src/main/java/org/apache/spark/network/util/RocksDBProvider.java:125]: @Ngone51 https://github.com/apache/spark/pull/37648 unified the use of `initDB` method. If it is merged, this method can be deleted.

**Review by mridulm — APPROVED** [review]
Looks good to me - this PR depends on other PR, so we should wait for those to be merged.

+CC @Ngone51

File: common/network-common/src/main/java/org/apache/spark/network/util/RocksDBProvider.java
Comment (Ngone51) [reply common/network-common/src/main/java/org/apache/spark/network/util/RocksDBProvider.java:81]: > I think it is safer to follow the old code flow, could LevelDBProvider and RocksDBProvider be fixed together after further investigation?

sounds good.

**Comment by mridulm** [comment]
Will wait for a day or so to given Dongjoon a chance to review the PR.
Would be great if you can review this PR @dongjoon-hyun, particularly `RocksDBProvider`. Thanks !

File: resource-managers/yarn/src/test/scala/org/apache/spark/network/yarn/YarnShuffleServiceSuite.scala
Comment (LuciferYang) [inline resource-managers/yarn/src/test/scala/org/apache/spark/network/yarn/YarnShuffleServiceSuite.scala:691]: will re-enable this due to it should pass after SPARK-40186 merged

```
mvn clean test -pl resource-managers/yarn -Pyarn -DwildcardSuites=org.apache.spark.network.yarn.YarnShuffleServiceWithRocksDBBackendSuite
```

```
YarnShuffleServiceWithRocksDBBackendSuite:
- executor and merged shuffle state kept across NM restart
- removed applications should not be in registered executor file and merged shuffle file
- shuffle service should be robust to corrupt registered executor file
- get correct recovery path
- moving recovery file from NM local dir to recovery path
- service throws error if cannot start
- Consistency in AppPathInfo between in-memory hashmap and the DB
- Finalized merged shuffle are written into DB and cleaned up after application stopped
- SPARK-40186: shuffleMergeManager should have been shutdown before db closed
- Dangling finalized merged partition info in DB will be removed during restart
- Dangling application path or shuffle information in DB will be removed during restart
- Cleanup for former attempts local path info should be triggered in applicationRemoved
- recovery db should not be created if NM recovery is not enabled
- SPARK-31646: metrics should be registered into Node Manager's metrics system
- SPARK-34828: metrics should be registered with configured name
- create default merged shuffle file manager instance
- create remote block push resolver instance
- invalid class name of merge manager will use noop instance
Run completed in 9 seconds, 454 milliseconds.
Total number of tests run: 18
Suites: completed 2, aborted 0
Tests: succeeded 18, failed 0, canceled 0, ignored 0, pending 0
All tests passed.
```

**Comment by LuciferYang** [comment]
[01220ed](https://github.com/apache/spark/pull/37610/commits/01220ed5a2256a7417c04ae6f67ecd5c7d011ae2) merge with master and [52c1c7e](https://github.com/apache/spark/pull/37610/commits/52c1c7e11b98784ab695afe64d5bb5e519a7fcd5) re-enable `Finalized merged shuffle are written into DB and cleaned up after application stopped` in `YarnShuffleServiceSuite`

**Comment by mridulm** [comment]
Merged to master.
Thanks for working on this @LuciferYang !
Thanks for the review @Ngone51 :-)

**Comment by LuciferYang** [comment]
thanks @mridulm @Ngone51

**Comment by dongjoon-hyun** [comment]
Thank you, @LuciferYang, @mridulm , @Ngone51 and @HyukjinKwon .

**Pull request closed without merging**
